### PR TITLE
[AIRFLOW-6128] Simplify plugins_manager and webserver plugin code

### PR DESCRIPTION
**Pull request by potiuk**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6128

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

While adding type annotations, it was apparent that the code for
plugin manager and webserver initialisation was rather convoluted
and complex. This change fixes it as follow up after untangling
cyclic dependnencies.

As a side effect - it turned out that the test plugin view
in airflow RBAC implementation was not working since 1.10.1 - it threw "render
is not an attribute" error - this has been fixed in this commit
as well. The Plugins/TestView now works as expected. See
https://stackoverflow.com/questions/53701030/airflow-plugins-rbac-enabled-blueprint-not-working?answertab=active#tab-top

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by potiuk** [comment]
While working on adding more type annotations I've found more cyclic dependencies in Plugins manager. I managed to untangle them in this commit as a follow-up after the "executors" untangling.

As result - we have much smaller airflow.__init__.py, and the dependencies are much more straightforward and not cyclic:

```
airflow -> PluginsManager -> [operators/hooks/macros/sensors] 
```

rather than

```
airflow -> [operators/hooks/macros/sensors]  -> PluginsManager -> [operators/hooks/macros/sensors] 
```

I also made most of the affected files pylint + mypy compliant.

File: airflow/config_templates/airflow_local_settings.py [hunk 1]
```diff
@@ -201,17 +199,17 @@
...
+        ELASTICSEARCH_JSON_FORMAT: str = conf.get('elasticsearch', 'JSON_FORMAT')
+        ELASTICSEARCH_JSON_FIELDS: str = conf.get('elasticsearch', 'JSON_FIELDS')
 
-        ELASTIC_REMOTE_HANDLERS = {
+        ELASTIC_REMOTE_HANDLERS: Dict[str, Dict[str, str]] = {
             'task': {
                 'class': 'airflow.utils.log.es_task_handler.ElasticsearchTaskHandler',
                 'formatter': 'airflow',
-                'base_log_folder': os.path.expanduser(BASE_LOG_FOLDER),
+                'base_log_folder': str(os.path.expanduser(BASE_LOG_FOLDER)),
```
Comment (ashb) [inline airflow/config_templates/airflow_local_settings.py]: The changes in this file seem to be just type hint related, so should be unrelated to this change?

**Comment by potiuk** [comment]
Yeh. I actually used type-hints and mypy to verify if my changes are correct (except running the airflow). It is super-helpful to be confident about your changes while commiting them. I am using pre-commit extensively to make sure I do not make mistakes. It's a big value of pylint/mypy/flake that I am using.

I will split out typehint changes as the last step of review. Similarly as I do for #6596 . But this has to come as next steps. First I use mypy/pylint to get my changes verified in automated way and then I can split them out to make sure they are change/bisect friendly.

**Comment by potiuk** [comment]
@ashb-> separated out type adding to #6140. Now what we are left with is a change that internally splits-out plugins_manager and webserver initialisation to be easier to understand - includes type annotations and fix to the long-standing problem of TestView bleupring not working since 1.10.1 https://stackoverflow.com/questions/53701030/airflow-plugins-rbac-enabled-blueprint-not-working?answertab=active#tab-top

File: airflow/config_templates/airflow_local_settings.py (hunk 1 above)
Comment (potiuk) [reply airflow/config_templates/airflow_local_settings.py]: I split it out to separate commit.

**Comment by potiuk** [comment]
@ashb, @kaxil @feluelle @Fokko @mik-laj  -> another spin-off after refactoring of executors. This time it's Plugins manager (look at the last commit only - I separated out adding type annotation to previous commit).

File: airflow/plugins_manager.py [hunk 2]
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
...
-        filepath = os.path.join(root, f)
-        try:
-            if not os.path.isfile(filepath):
-                continue
-            mod_name, file_ext = os.path.splitext(
-                os.path.split(filepath)[-1])
-            if file_ext != '.py':
-                continue
+def import_plugin(filepath: str, mod_name: str, root: str) -> None:
+    """Imports plugin module."""
```
Comment (feluelle) [inline airflow/plugins_manager.py:171]: Don't you think a full documentation is needed here?

File: airflow/plugins_manager.py [hunk 3]
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
...
-            for obj in list(m.__dict__.values()):
-                if is_valid_plugin(obj, plugins):
-                    plugins.append(obj)
-        except Exception as e:  # pylint: disable=broad-except
-            log.exception(e)
-            path = filepath or str(f)
-            log.error('Failed to import plugin %s', path)
-            import_errors[path] = str(e)
+def try_to_import_plugin(filepath: str, root: str) -> None:
+    """Tries to import plugin from the file specified. Skips it if not a valid python file."""
```
Comment (feluelle) [inline airflow/plugins_manager.py:183]: Same here.

File: airflow/plugins_manager.py [hunk 4]
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
...
+    import_plugin(filepath=filepath, mod_name=mod_name, root=root)
 
-plugins = load_entrypoint_plugins(
-    pkg_resources.iter_entry_points('airflow.plugins'),
-    plugins
-)
+
+def find_airflow_plugins() -> None:
+    """Crawl through the plugins folder to find AirflowPlugin derivatives"""
+    assert settings.PLUGINS_FOLDER, "Plugins folder is not set"
```
Comment (feluelle) [inline airflow/plugins_manager.py:194]: I thought we did decide to not use `assert` for raising exceptions.

File: airflow/plugins_manager.py [hunk 5]
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
...
-    })
 
-if len(stat_name_handlers) > 1:
-    raise AirflowPluginException(
-        'Specified more than one stat_name_handler ({}) '
-        'is not allowed.'.format(stat_name_handlers))
+def get_stat_name_handler() -> Optional[Callable[[str], str]]:
+    """Retrieves the single configured stats_handler (there can be only one!)"""
+    assert len(stat_name_handlers) in {0, 1}, \
+        f'Specified more than one stat_name_handler ({stat_name_handlers}) which is not allowed.'
```
Comment (feluelle) [inline airflow/plugins_manager.py:288]: > I thought we did decide to not use assert for raising exceptions.

Same here.

File: airflow/plugins_manager.py [hunk 6]
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
...
-                       link.__class__.__name__): link.__class__
-        for link in p.operator_extra_links
-    })
 
-if len(stat_name_handlers) > 1:
-    raise AirflowPluginException(
-        'Specified more than one stat_name_handler ({}) '
-        'is not allowed.'.format(stat_name_handlers))
+def get_stat_name_handler() -> Optional[Callable[[str], str]]:
+    """Retrieves the single configured stats_handler (there can be only one!)"""
```
Comment (feluelle) [inline airflow/plugins_manager.py:286]: Can you add the documentation for the return value, please?

File: airflow/www/app.py [hunk 7]
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
...
 
-            return globals
 
-        @app.teardown_appcontext
-        def shutdown_session(exception=None):  # pylint: disable=unused-variable
-            settings.Session.remove()
-
-    return app, appbuilder
+def configure_app(app_name: str, config: Optional[Dict[str, Union[str, bool]]], testing: bool):
+    """Configure the application."""
```
Comment (feluelle) [inline airflow/www/app.py:242]: Please add a description to the parameters.

File: airflow/www/app.py [hunk 8]
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
...
+    app.config['SESSION_COOKIE_HTTPONLY'] = True
+    app.config['SESSION_COOKIE_SECURE'] = conf.getboolean('webserver', 'COOKIE_SECURE')
+    app.config['SESSION_COOKIE_SAMESITE'] = conf.get('webserver', 'COOKIE_SAMESITE')
+    if config:
+        app.config.from_mapping(config)
 
 
-def root_app(env, resp):
+def root_app(_, resp: Callable[[bytes, List[Tuple[str, str]]], None]):
+    """Root application function - returns 404 if the path is not found."""
```
Comment (feluelle) [inline airflow/www/app.py:266]: Same here.

File: airflow/www/app.py [hunk 9]
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
...
     resp(b'404 Not Found', [('Content-Type', 'text/plain')])
     return [b'Apache Airflow is not at this location']
 
 
-def cached_app(config=None, session=None, testing=False):
-    global app, appbuilder
+def cached_app(config: Optional[Dict[str, Union[str, bool]]] = None,
+               session: Optional[Session] = None,
+               testing: bool = False) -> Flask:
+    """Returns cached application if it's there. Creates it otherwise."""
```
Comment (feluelle) [inline airflow/www/app.py:274]: And here.

File: airflow/www/app.py [hunk 10]
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
...
+        if not app:
+            raise AirflowException("The app cannot be None")
     return app
 
 
-def cached_appbuilder(config=None, testing=False):
-    global appbuilder
+def cached_appbuilder(config: Optional[Dict[str, Union[str, bool]]] = None, testing=False) -> \
+        AppBuilder:
+    """Returns cached appbuilder."""
```
Comment (feluelle) [inline airflow/www/app.py:289]: And here.

**Comment by potiuk** [comment]
Thanks @feluelle ! I will come back to it soon-ish (or late-ish maybe)

**Comment by potiuk** [comment]
Same story - I think I will close that one. Ane refactoring/removals of cyclic dependencies etc. will be far easier when we move things outside of the airflow.__init__.py

**Pull request closed without merging**